tests/group_functions/type_coverage.h

```diff
@@ -0,0 +1,708 @@
+ */
+
+template <typename T, typename X, typename... Rs>
+static constexpr bool present_in = present_in<T, Rs...>;
+
+template <typename T, typename X, typename... Rs>
+static constexpr bool present_in<T, T, X, Rs...> = true;
+
+template <typename T, typename X>
+static constexpr bool present_in<T, X> = std::is_same_v<T, X>;
+
+///////////////////////////////////////////////////////////////////////////////
+
+/**
+ * @brief Overload to handle cases where no runtime arguments provided with
+ * unnamed type packs
+ */
+template <template <typename...> class Action, typename Needed,
+          typename... ActionArgsT>
+inline void for_all_combinations_with() {
+  if constexpr (present_in<Needed, ActionArgsT...>) Action<ActionArgsT...>{}();
+}
+
+/**
+ * @brief Generic function to run specific action for every combination of each
+ * of the types given by appropriate type pack instances with one type
+ * needed to be present in any of the instances.
+ * Virtually any combination of named and unnamed type packs is supported.
+ * Supports different types of compile-time value lists via value pack.
+ * @tparam Action Functor template for action to run
+ * @tparam Needed Type needed to be present inside the functor template
+ * instantiation
+ * @tparam ActionArgsT Parameter pack to use for functor template instantiation
+ * @tparam HeadT The type of the first non-pack argument during the recursion
+ * @tparam ArgsT Parameter pack with types of arguments for functor
+ * @param head The first non-pack argument to pass into the functor
+ * @param args The rest of the arguments to pass into the functor
+ */
+template <template <typename...> class Action, typename Needed,
+          typename... ActionArgsT, typename HeadT, typename... ArgsT,
+          sfinae::is_not_a_type_pack<HeadT> = true>
+inline void for_all_combinations_with(HeadT &&head, ArgsT &&...args) {
+  // The first non-pack argument passed into the for_all_combinations stops the
+  // recursion
+  if constexpr (present_in<Needed, ActionArgsT...>)
+    Action<ActionArgsT...>{}(std::forward<HeadT>(head),
+                             std::forward<ArgsT>(args)...);
+}
+
+/**
+ * @brief Overload to handle the iteration over the types within the named type
+ * pack
+ */
+template <template <typename...> class Action, typename Needed,
+          typename... ActionArgsT, typename... HeadTypes, typename... ArgsT>
+inline void for_all_combinations_with(const named_type_pack<HeadTypes...> &head,
+                                      ArgsT &&...args) {
+  // Run the next level of recursion for each type from the head named_type_pack
+  // instance. Each recursion level unfolds the first argument passed and adds a
+  // type name as the last argument.
+  size_t type_name_index = 0;
+
+  ((for_all_combinations_with<Action, Needed, ActionArgsT..., HeadTypes>(
+        std::forward<ArgsT>(args)..., head.names[type_name_index]),
+    ++type_name_index),
```

Comment:
It looks similar to the type product library in an other file earlier.
Any way to factorize?
All this seems quite complex...

Reply:
I wonder whether we should not use a meta-programming library from Boost.

Reply:
I usually prefer self-contained solutions, as well as the authors of previous tests (from `tests/common/type_coverage.h`). It is possible to rewrite this using Boost, but then many tests outside of the STC contract scope should be refactored.